File: problems/46/solution3.py
# Accepted: 38ms (83.13%), 16.90MB (54.90%)

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        permutations: List[List[int]] = []
        def go(k: int = len(nums)) -> None:
            if k == 1:
                permutations.append(nums[:])
                return

            go(k-1)

            for i in range(k-1):
                if k % 2 == 0:
                    nums[i], nums[k-1] = nums[k-1], nums[i]
                else:
                    nums[0], nums[k-1] = nums[k-1], nums[0]

                go(k-1)
        go()

        return permutations

File: problems/46/solution3.cpp
// Accepted: 0ms (100.00%), 10.51MB (5.18%)

class Solution {
    vector<vector<int>> permutations;

    void go(int k, vector<int>& nums) {
        if (k == 1) {
            vector<int> permutation(nums.begin(), nums.end());
            permutations.push_back(permutation);
            return;
        }

        go(k-1, nums);
        for (int i = 0; i < k-1; ++i) {
            if (k % 2 == 0) {
                swap(nums[i], nums[k-1]);
            } else {
                swap(nums[0], nums[k-1]);
            }
            go(k-1, nums);
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        go(nums.size(), nums);

        return permutations;
    }
};
